File: dimdwarf-core/src/main/java/net/orfjackal/dimdwarf/net/SimpleSgsProtocolEncoder.java
// Copyright © 2008-2010 Esko Luontola <www.orfjackal.net>
// This software is released under the Apache License 2.0.
// The license text is at http://dimdwarf.sourceforge.net/LICENSE

package net.orfjackal.dimdwarf.net;

import com.sun.sgs.protocol.simple.SimpleSgsProtocol;
import org.apache.mina.core.buffer.IoBuffer;
import org.apache.mina.core.session.IoSession;
import org.apache.mina.filter.codec.*;

import javax.annotation.concurrent.Immutable;

/**
 * Encodes {@link com.sun.sgs.protocol.simple.SimpleSgsProtocol} messages.
 */
@Immutable
public class SimpleSgsProtocolEncoder implements ProtocolEncoder {

    // TODO: ensure compatibility with com.sun.sgs.impl.sharedutil.MessageBuffer

    public void encode(IoSession session, Object message, ProtocolEncoderOutput out) throws Exception {
        if (message instanceof LoginFailure) {
            IoBuffer buffer = IoBuffer.allocate(5);
            buffer.putShort((short) 3); // payload length
            buffer.put(SimpleSgsProtocol.LOGIN_FAILURE);
            buffer.putShort((short) 0); // (String) message for login failure (length = 0)
            buffer.flip();
            out.write(buffer);
        }
    }

    public void dispose(IoSession session) throws Exception {
    }
}

File: dimdwarf-core/src/main/scala/net/orfjackal/dimdwarf/net/SimpleSgsProtocolEncoder.scala
package net.orfjackal.dimdwarf.net

import com.sun.sgs.protocol.simple.SimpleSgsProtocol
import org.apache.mina.filter.codec._
import org.apache.mina.core.session.IoSession
import org.apache.mina.core.buffer.IoBuffer
import javax.annotation.concurrent.Immutable

@Immutable
class SimpleSgsProtocolEncoder extends ProtocolEncoder {
  def encode(session: IoSession, message: Any, out: ProtocolEncoderOutput) {
    val encoded = message match {

      case LoginSuccess() =>
        IoBuffer.allocate(5).
                putShort(3.asInstanceOf[Short]).
                put(SimpleSgsProtocol.LOGIN_SUCCESS).
                putShort(0.asInstanceOf[Short]).
                flip()

      case LoginFailure() =>
        IoBuffer.allocate(5).
                putShort(3.asInstanceOf[Short]).
                put(SimpleSgsProtocol.LOGIN_FAILURE).
                putShort(0.asInstanceOf[Short]).
                flip()

      case LogoutSuccess() =>
        IoBuffer.allocate(3).
                putShort(1.asInstanceOf[Short]).
                put(SimpleSgsProtocol.LOGIN_SUCCESS).
                flip()
    }
    out.write(encoded)
  }

  def dispose(session: IoSession) {
  }
}
